stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/observability/WebSocketObservabilityUtil.java

```diff
@@ -0,0 +1,237 @@
+        } else {
+            logger.debug("WebSocket type:{}/{}, message: {}, connectionId: {}, service/url: {}",
+                         errorType, messageType, errorMessage, observerContext.getConnectionId(),
+                         observerContext.getServicePathOrClientUrl());
+        }
+    }
+
+    /**
+     * Observes WebSocket errors before the connection is initialized (handshake errors etc).
+     *
+     * @param errorType      type of error (connection, closure, message sent/received).
+     * @param errorMessage   error message.
+     * @param url            service base path or remote url.
+     * @param clientOrServer current context of connection.
+     */
+    public static void observeError(String errorType, String errorMessage, String url, String clientOrServer) {
+        WebSocketMetricsUtil.reportErrorMetrics(errorType, url, clientOrServer);
+
+        logger.debug("WebSocket type:{}, message: {}, service/url: {}", errorType, errorMessage, url);
+    }
+
+    /**
+     * Observes when a new resource is invoked. In addition to metrics and logging, relevant tags are added to the
+     * span in the trace as well.
+     *
+     * @param strand            current strand
+     * @param connectionInfo    information regarding the connection.
+     * @param resource          name of the resource invoked.
+     */
+    public static void observeResourceInvocation(Strand strand, WebSocketConnectionInfo connectionInfo,
+                                                 String resource) {
+        WebSocketTracingUtil.traceResourceInvocation(strand, connectionInfo);
+        observeResourceInvocation(connectionInfo, resource);
+    }
+
+    /**
+     * Observes when a new resource is invoked.
+     *
+     * @param connectionInfo    information regarding the connection.
+     * @param resource          name of the resource invoked.
+     */
+    public static void observeResourceInvocation(WebSocketConnectionInfo connectionInfo, String resource) {
+        WebSocketObserverContext observerContext = new WebSocketObserverContext(connectionInfo);
+        WebSocketMetricsUtil.reportResourceInvocationMetrics(observerContext, resource);
+
+        logger.debug("WebSocket resource invoked. connectionID: {}, service/url: {}, resource: {}",
+                     observerContext.getConnectionId(),
+                     observerContext.getServicePathOrClientUrl(),
+                     resource);
+    }
+
+    /**
+     * Determines whether the current context is server or client and sets the tags accordingly. If server, the service
+     * base path. If client, the remote URL
+     *
+     * @param connectionInfo information regarding connection.
+     * @return whether the current context is client or server.
+     */
+    static String getClientOrServerContext(WebSocketConnectionInfo connectionInfo) {
+        WebSocketService service = connectionInfo.getService();
+        //if the current context is server
```

Comment:
```suggestion
```

These comments are quite obvious. You need comments when the code might be a little too complicated to understand.